File: Medium/1726-Tuple with Same Product/HashMap.rb
# @param {Integer[]} nums
# @return {Integer}
def tuple_same_product(nums)
  counter = {}
  counter.default = 0

  (0...nums.size).each do |i|
    (i + 1...nums.size).each do |j|
      counter[nums[i] * nums[j]] += 1
    end
  end

  counter.values.map { |c| c * (c - 1) * 4 }.sum
end

File: Problemset/1726-Tuple with Same Product/HashMap.rs
use std::collections::HashMap;

impl Solution {
    pub fn tuple_same_product(nums: Vec<i32>) -> i32 {
        let mut counter = HashMap::new();

        for i in 0..nums.len() {
            for j in (i + 1)..nums.len() {
                *counter.entry(nums[i] * nums[j]).or_insert(0) += 1;
            }
        }

        counter.values().map(|c| c * (c - 1) * 4).sum()
    }
}
